google-cloud-storage/src/test/java/com/google/cloud/storage/spi/v1/HttpStorageRpcTest.java

```diff
@@ -0,0 +1,313 @@
+            @Override
+            protected LowLevelHttpRequest buildRequest(String method, String url)
+                throws IOException {
+              if (TEST_MOCK_DATA.isEmpty()) {
+                throw new IllegalStateException("No test data provided");
+              }
+              MockData data = TEST_MOCK_DATA.poll();
+              assertEquals(data.method, method);
+              assertEquals(data.url, url);
+              TestRequest testRequest = new TestRequest(data.response);
+              data.rpcRequestHolder.setRequest(testRequest);
+              return testRequest;
+            }
+          };
+        }
+      };
+
+  private static final HttpTransportOptions TRANSPORT_OPTIONS =
+      HttpTransportOptions.newBuilder().setHttpTransportFactory(TRANSPORT_FACTORY_MOCK).build();
+
+  private static final StorageOptions STORAGE_OPTIONS =
+      StorageOptions.newBuilder()
+          .setProjectId("projectId")
+          .setClock(TIME_SOURCE)
+          .setServiceRpcFactory(RPC_FACTORY_MOCK)
+          .setRetrySettings(ServiceOptions.getNoRetrySettings())
+          .setTransportOptions(TRANSPORT_OPTIONS)
+          .build();
+
+  static class RpcRequestHolder {
+    TestRequest request = null;
+
+    void setRequest(TestRequest request) {
+      this.request = request;
+    }
+
+    TestRequest getRequest() {
+      return request;
+    }
+  }
+
+  static class MockData {
+    final String method;
+    final String url;
+    final LowLevelHttpResponse response;
+    final RpcRequestHolder rpcRequestHolder;
+
+    MockData(
+        String method, String url, LowLevelHttpResponse response, RpcRequestHolder requestHeaders) {
+      this.method = method;
+      this.url = url;
+      this.response = response;
+      this.rpcRequestHolder = requestHeaders;
+    }
+  }
+
+  /**
+   * Adds mock parameters for the RPC call.
+   *
+   * @param method expected HTTP method
+   * @param url expected URL
+   * @param response HTTP response object to be returned
+   * @return a holder object which will contain the Http Request made by RPC
+   */
+  private static RpcRequestHolder mockResponse(
+      String method, String url, LowLevelHttpResponse response) {
+    RpcRequestHolder holder = new RpcRequestHolder();
+    TEST_MOCK_DATA.add(new MockData(method, url, response, holder));
+    return holder;
+  }
+
+  private static String getUnzippedContent(LowLevelHttpRequest request) {
+    try {
+      ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+      request.getStreamingContent().writeTo(outputStream);
+      GZIPInputStream gzipInputStream =
+          new GZIPInputStream(new ByteArrayInputStream(outputStream.toByteArray()));
+      return new String(ByteStreams.toByteArray(gzipInputStream), UTF_8);
+    } catch (IOException e) {
+      throw new Error(e);
+    }
+  }
+
+  @Before
+  public void setUp() throws Exception {
+    rpc = new HttpStorageRpc(STORAGE_OPTIONS);
+    TEST_MOCK_DATA.clear();
```

Comment:
yes, this really should be an instance variable so you don't have to worry about clearing it

Reply:
refactored to be an instance variable